src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    ffi::CString,
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use wayland_protocols_misc::zwp_virtual_keyboard_v1::server::zwp_virtual_keyboard_v1::{
+    self, ZwpVirtualKeyboardV1,
+};
+use wayland_server::{
+    backend::{ClientId, ObjectId},
+    protocol::wl_keyboard::{KeyState, KeymapFormat},
+    Client, DataInit, Dispatch, DisplayHandle, Resource,
+};
+use xkbcommon::xkb;
+
+use crate::{
+    input::{keyboard::KeymapFile, Seat, SeatHandler},
+    utils::SERIAL_COUNTER,
+    wayland::seat::{keyboard::for_each_focused_kbds, WaylandFocus},
+};
+
+use super::VirtualKeyboardManagerState;
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Hash)]
+struct SerializedMods {
+    depressed: u32,
+    latched: u32,
+    locked: u32,
+    group: u32,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct VirtualKeyboard {
+    pub instances: Vec<ZwpVirtualKeyboardV1>,
+    modifiers: Option<SerializedMods>,
+    old_keymap: Option<KeymapFile>,
+}
+
+/// Handle to a virtual keyboard instance
+#[derive(Debug, Clone, Default)]
+pub(crate) struct VirtualKeyboardHandle {
+    pub(crate) inner: Arc<Mutex<VirtualKeyboard>>,
+}
+
+impl VirtualKeyboardHandle {
+    pub(super) fn add_instance<D>(&self, instance: &ZwpVirtualKeyboardV1) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.instances.push(instance.clone());
+    }
+}
+
+/// User data of ZwpVirtualKeyboardV1 object
+#[derive(Debug)]
+pub struct VirtualKeyboardUserData<D: SeatHandler> {
+    pub(super) handle: VirtualKeyboardHandle,
+    pub(crate) seat: Seat<D>,
+}
+
+impl<D> Dispatch<ZwpVirtualKeyboardV1, VirtualKeyboardUserData<D>, D> for VirtualKeyboardManagerState
```

Comment:
- A ZwpVirtualKeyboardV1 has a strong reference to the VirtualKeyboardUserData.
- A VirtualKeyboardUserData contains a `VirtualKeyboard`
- A VirtualKeyboard contains a `Vec<ZwpVirtualKeyboardV1>`

I think it's pretty likely we have a reference cycle which will leak memory.

Reply:
VirtualKeyboard now contains just a u8, I wasn't using the reference for anything anyway, just needed to count virtual keyboard instances.